[RFC] jobs: guarantee unique job IDs

splinterofchaos opened the pull request:
From the commit message:

Script writers assume that job IDs uniquely identify them, but using
their index in the job table can break this assumption because the
indices will be re-used.

~~Process IDs make ideal, unique identifiers, so store jobs in a map with
their PID as they key, and also contains more useful information for
users.~~ An increment only counter is used instead.

fixes #2162

Commit:
jobs: guarantee (almost) uniqueness for job IDs

Script writers assume that job IDs uniquely identify them, but using
their index in the job table can break this assumption because the
indices will be re-used.

Use an increment-only counter between 1 and INT_MAX (inclusive) for
assigning job IDs to make aliasing highly unlikely. (Running out of
system memory is more likely than having INT_MAX jobs.)

fixes #2162

ZyX-I commented:
Process IDs are being constantly reused as well. They are only unique identifiers for *currently running* processes, not for processes launched since OS started. Check out `cat /proc/sys/kernel/pid_max`: on my system this is only 32768. Nothing compared to amount of launched processes given shell scripts used for many jobs (which start process on almost every action).

If you do need unique identifiers you need to have a something similar to code used for buffers (it is linked list, but buffer number is obtained using incremented-only global counter).

Other then that I think that PR is worth merging only because it gets rid of MAX_RUNNING_JOBS. But commit message needs to be changed because process ID cannot be “ideal, unique identifier” that is never reused.

ZyX-I commented:
Ah, and that will not fix #2162.

splinterofchaos commented:
> If you do need unique identifiers you need to have a something similar to code used for buffers (it is linked list, but buffer number is obtained using incremented-only global counter).

Dang, I thought that *was* the standard algorithm for producing PIDs, but apparently Windows, for example, offers no such guarantee. Oh well, I repushed with that solution.

http://blogs.msdn.com/b/oldnewthing/archive/2011/01/07/10112755.aspx

oni-link commented on src/nvim/os/job.c:
```diff
@@ -55,6 +56,22 @@ uv_signal_t schld;
+/// for two jobs to have the same ID.
+static int new_job_id(void)
+{
+  count = count < INT_MAX ? count + 1 : 1;
```
Is `0` a special value for `count`? Otherwise we could use an unsigned type for `count` and get the cyclical counter with just `count++`.

ZyX-I replied:
@oni-link: From `:h jobstart`:

>		Returns:
>		  - The job ID on success, which is used by |jobsend()| and
>		    |jobstop()|
>		  - 0 when the job table is full or on invalid arguments
>		  - -1 when {cmd}[0] is not executable.  Will never fail if
>		    {cmd} is a string unless 'shell' is not executable.

. 0 and -1 are reserved. If you simply use `unsigned` then you will be able to have *both* values because in VimL there are only signed integers.

ZyX-I replied:
By the way, now nobody checks for the condition “job table is full”: this is now equivalent to an infinite cycle.

splinterofchaos replied:
> By the way, now nobody checks for the condition “job table is full”

fixed

ZyX-I commented:
AFAIK process IDs are obtained using incremented counter that wraps (and, of course, with skipping of PIDs already in use), except for some systems. But in any case I know *no* systems where PIDs are unique or maximum PID has magnitude similar to INT_MAX.

splinterofchaos commented:
> But in any case I know no systems where PIDs are unique 

Including this one! I don't think it's possible to ensure absolute uniqueness in a robust system, so I only want to promise "unique within reason".

ZyX-I commented on src/nvim/os/job.c:
```diff
@@ -62,20 +83,22 @@ void job_init(void)
-    }
-  }
+  Job *job;
+  int id; (void)id;
```
Wondering why there is `(void)id`. Can you leave a comment (I mean, leave in a *source code*) for future reference?

ZyX-I replied:
I have found the same construct in the only other occurence of `map_foreach`, but removing it does not add any warnings. And this is not documented either.

oni-link replied:
It prevents a compiler warning in Release mode.
```bash
/home/oni-link/git/neovim/src/nvim/os/job.c: In function ‘job_teardown’:
/home/oni-link/git/neovim/src/nvim/os/job.c:95:7: warning: variable ‘id’ set but not used [-Wunused-but-set-variable]
   int id; //(void)id;
       ^
```

ZyX-I replied:
There is `map_foreach_value` for this case; “the only other occurrence in `terminal.c`” happened to iterate over keys (there is no `map_foreach_key` for some reason).

splinterofchaos replied:
Thanks for the tip about `map_foreach_value`!

> Wondering why there is (void)id.

I also ~~get~~ got this warning with debug mode.

ZyX-I commented on src/nvim/os/job.c:
```diff
@@ -92,27 +115,15 @@ void job_teardown(void)
+  // Initialize
+  job->id = new_job_id();
+  *status = job->id;
+  if (*status == -1) {
```
Should not it do `xfree(job)` here?

splinterofchaos replied:
:+1: -- actually, if we can't allocate a job ID, might as well not `xmalloc()` to begin with.

ZyX-I commented on src/nvim/os/job.c:
```diff
@@ -381,11 +390,12 @@ JobOptions *job_opts(Job *job)
-    if ((job = table[i]) == NULL || !job->stopped_time) {
+  Job *job;
+  int id; (void)id;
+  map_foreach(job_map, id, job, {
```
map_foreach_value

splinterofchaos replied:
:+1:

ZyX-I commented on src/nvim/os/job.c:
```diff
@@ -450,23 +460,30 @@ static void chld_handler(uv_signal_t *handle, int signum)
+  Job *job;
+  bool found = false;
+  int id; (void)id;
+  map_foreach(job_map, id, job, {
```
map_foreach_value

splinterofchaos replied:
:+1:

oni-link commented on src/nvim/os/job.c:
```diff
@@ -175,8 +185,8 @@ Job *job_start(JobOptions opts, int *status)
-  // Save the job to the table
-  table[i] = job;
+
+  pmap_put(int)(job_map, job->id, job);
```
Looks like `*status == 0` is not needed anymore as an error.

splinterofchaos replied:
Oops, I got confused and passed `-1` for the "Job table full" error... although, really, I think that will never happen.

oni-link replied:
Would it be better to remove this error completely? On linux we can only have ~32k processes.

ZyX-I replied:
@oni-link On linux you may have up to 2²² processes. ≈32k is the default limit, but it can be increased.

ZyX-I replied:
@oni-link You may use `echo 4194304 > /proc/sys/kernel/pid_max` to do this (sets max PID to 2²²). AFAIK this applies to 64-bit systems only, 32-bit have maximum limit 32768 (see `man 5 proc`, paragraph regarding pid_max file).

oni-link commented on src/nvim/os/job.c:
```diff
@@ -92,27 +115,14 @@ void job_teardown(void)
-    }
+  *status = new_job_id();
+  if (*status == 0) {
+    return NULL;  // We somehow allocated MAX_INT job IDs!
```
`shell_free_argv(opts.argv);` is missing.

splinterofchaos replied:
:+1: -- thanks

splinterofchaos commented:
I'd like to mark this RDY, but it looks like I haven't gotten a LGTM, yet. Any further comments?

Sorry for letting this go inactive. Got involved in a new project.

oni-link commented:
The error message for a failed spawn could be changed. The following script gives error `E902: "%s" is not an executable` (too many files open?) instead of `E903: Could not spawn API job`:
```vim
while 1
  call jobstart("sleep 10")
endw
```

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 6bd2b23..3c2692d 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -20262,7 +20262,7 @@ static inline Job *common_job_start(JobOptions opts, typval_T *rettv)
       xfree(opts.term_name);
       free_term_job_data(data);
     } else {
-      EMSG(_(e_jobexe));
+      EMSG(_(e_api_spawn_failed));
     }
     return NULL;
   }
```
Otherwise, LGTM.

Commit:
jobstart: improve error message

splinterofchaos commented:
> The following script gives error E902: "%s" is not an executable (too many files open?)

Yes, and we currently ignore `errno`, which would tell us that, so I added a commit to have `e_jobexe` tell us `"Could not spawn "sh": too many open files"`

Pyrohh commented:
@justinmk perhaps this should be tagged in a near-future milestone?

splinterofchaos commented on src/nvim/os/pty_process.c:
```diff
@@ -113,31 +114,34 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
-  if (job->opts.stderr_cb
-      && !set_pipe_duplicating_descriptor(master, &ptyproc->proc_stderr)) {
+  if (job->opts.stderr_cb &&
+      (err = set_pipe_duplicating_descriptor(master, &ptyproc->proc_stderr))) {
```
These `if`s do break the style, but in order to keep the line under 80 chars, I either had to break after the &&` or put the `{` on the next line.

ZyX-I replied:
I would in this case put &ptyproc… argument on the next line, making it three lines total.

splinterofchaos commented:
Sorry for the very long delay.

@oni-link's comments seem to have disappeared, but I fixed the error return of `pty_process.c` and updated the docs, and rebased.

oni-link commented on src/nvim/os/pty_process.c:
```diff
@@ -99,11 +99,12 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
 
   if (pid < 0) {
-    return false;
+    return pid;
```
Why not return `-errno` like the other error code paths.

oni-link commented on src/nvim/eval.c:
```diff
@@ -20263,10 +20265,12 @@ static inline Job *common_job_start(JobOptions opts, typval_T *rettv)
       free_term_job_data(data);
     } else {
-      EMSG(_(e_jobexe));
+      EMSG3(_(e_jobexe), progname, uv_strerror(rettv->vval.v_number));
```
The third argument here is always `-1` and is not the uv error. The value returned in case of a failed spawn could be used as error status from `job_start()`.

oni-link commented on src/nvim/os/job.c:
```diff
@@ -141,7 +152,8 @@ Job *job_start(JobOptions opts, int *status)
 
   // Spawn the job
-  if (!process_spawn(job)) {
+  int result = process_spawn(job);
```
If we don't use this as value for `*status` , we could remove the variable.

tarruda commented:
@splinterofchaos in #2980 the job/process modules have suffered great changes and now each job has a unique id, so this PR is no longer  necessary.

Thanks anyway
